runtime/middlewares.go

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"context"
+	"net/http"
+)
+
+// Handler is an interface that compatable middlewares must implement to be used
+// in the middleware stack. Handle should yield to the next middleware in the stack
+// by invoking the next HandlerFn passed in.
+type Handler interface {
```

Comment:
This middleware interface is not very performant or debuggable.

We do not want people to author middlewares as closures. Generally writing closures instead of structs with methods is less performant and harder to debug because state is stored as closed over variables instead of fields on structs.

Having a wrapping style middleware makes reasoning about the performance profile and flamegraphs difficult.

Currently our flamegraphs look something like this :

![image](https://cloud.githubusercontent.com/assets/479538/24311276/ae793e6a-1090-11e7-8b8e-b822cc445c7a.png)

If we wrap each middleware as a closure like this each middleware will add at least one new stack frame, eventually it will look something like this :

![image](https://cloud.githubusercontent.com/assets/479538/24311373/1ece34d6-1091-11e7-9fec-b994086f235c.png)

We want to avoid all of this noise in the stack frames so it's best if middlewares were a struct with a HandleRequest() / HandleResponse() interface. We then need a middleware runner that loops over the middlewares and calls them one by one and then finally calls the endpoint once all the middlewares are done.